### ISSUE TITLE
Support for .NetCore

### ISSUE THREAD

**Issue by davzucky** [empira/PDFsharp-1.5#6]
I'm starting to use your library to parse pdf and extract data from it. This is working perfectly for windows, however I'm not able to even compile with .Net core because of all the GUI library ( like silverlight or Winform) dependency that the base library has.
Do you thing it would be possible to have a base pdfSharp.Core library that only contain parsing and creation of of pdf ?
I'm would be happy to help if you need

**Comment by shontykov** [comment]
I havent bumped to .Net.core yet. But, yes it would be an amazing idea to
have pdfsharp.core library.

**Comment by daniellor** [comment]
my fork working for mono
https://github.com/daniellor/PDFsharp

**Comment by ThomasHoevel** [comment]
PDFsharp uses the GUI for image handling. We have plans to handle images in own code, but this is not going to happen soon. Creating PDFs without images is of little use in most cases.
Parsing PDFs is a different story. This should be rather simple with .NET Core.
I don't know if Stefan has any plans to support .NET Core.

**Comment by charabaruk** [comment]
It would definitely be very helpful to remove dependencies on GDI+ and WPF for PDF generation.

**Comment by andekande** [comment]
There is already a conditional build flag #Core spread in the code. I don't assume its specifically for .Net Core but a first step to get what we want. Even if Image handling won't work - we could use Vectors instead. Sufficient for creating invoices and reports.
There is currently no dependency-less solution available for creating PDFs in .Net Core ! This one has a great potential.
I also offer help. Glad to hear back from the maintainers if we can count on theirs support.

**Comment by ThomasHoevel** [comment]
There already is a CORE flag with a different meaning from a time before .NET Core was announced. The CORE build is not yet fully implemented and still has some dependencies on GDI+.

**Comment by finnsea** [comment]
Has anyone been working on a port that is usable in a UWP application? We have a UWP app for one of our clients that creates PDF documents using Component One that need to be merged before printing and uploading to a server. Component One unfortunately for us does not have the functionality to merge existing PDFs.

**Comment by groege** [comment]
I've started on a port 
https://github.com/groege/PdfSharpCore
Though I have to warn you that it works for me in my case for creating a delivery note pdf, it might not work for you - but you can try anyway ;)
Examples are also not up to date - but if you want to try it out you can always create an issue ;)

**Comment by o-leary** [comment]
Really interested in this. I'm running VS 2015 and trying to get PDFsharp working with a UWP app, how do we include your code?

**Comment by o-leary** [comment]
Any chance you could share a built version? I can't get this to compile in VS 2015.

**Comment by groege** [comment]
Ah yes - I'm using VS2017 and new C# features ;)
https://github.com/groege/PdfSharpCore/blob/master/Example/PdfSharpCore.7z

Be aware of changes which I didn't include in the examples:
For images you must define the source by (i.e. Android):
ImageSource.ImageSourceImpl = new AndroidImageSource();
(Sadly I don't know how I can set it up to automatically use the right implementation)

For Fonts look at the example project.

Hope this helps ;)

**Comment by o-leary** [comment]
Thank you, I have got it working with a UWP app now, so far just simple text to test it. I'm sure I'll be able to figure the rest out on my own. Thank you very much for your help.

**Comment by groege** [comment]
No problem - let me know if you run into any issues ;)

**Comment by groege** [comment]
Actually i just remembered  - if you install the latest nuget package "Microsoft.Net.Compilers"  on the projects you should be able to compile it yourself ;)

**Comment by guylando** [comment]
It is not clear to me from this discussion if the original pdfsharp library is considering supporting .net core or they rely on groege to maintain the net core version? Similar questions were referenced here (https://github.com/empira/PDFsharp/issues/9) but there is no clear answer here to this question.

In my company, we want to use a stable well maintained library and if this library will not support net core we will have to search an alternative which is undesired.

**Comment by ThomasHoevel** [comment]
I do not know when an official PDFsharp version with support for .NET Core will be released. Latest I heard is that .NET Core 2.0 will greatly simplify the task. We did not talk about the ETA. If you need a library now, maybe try the groege port or other branches until the official version will be available.

**Comment by MiniverCheevy** [comment]
@groege any plans to push a version to nuget?

**Comment by ronnyek** [comment]
So dotnetcore2.0 stuff rolled out, but dont think its necessarily going to help/hurt a ton in the porting efforts. Nothing really changes with System.Drawing (if you get it in some form of compatibility, its only going to work in windows). That being said I've used ImageSharp stuff in dotnetcore projects, and its actually worked remarkably well.

**Comment by flensrocker** [comment]
FYI: There's now a compat-package for a lot of stuff:
https://www.nuget.org/packages/Microsoft.Windows.Compatibility/

And also a replacement for System.Drawing:
https://www.nuget.org/packages/System.Drawing.Common/

These can help on migrating PdfSharp to .Net Standard.

**Comment by StefanoD** [comment]
I have interest in a cross platform solution, too.

**Comment by tudorserbanconnatix** [comment]
Any progress with this? Version 1.50.4740-beta5 seems to be getting close?

**Comment by ronnyek** [comment]
I too am interested in this one...

**Comment by XelaNimed** [comment]
Is the implementation planned or not?

**Comment by ThomasHoevel** [comment]
AFAIK there are plans to create a version that compiles against .NET Standard and runs under Windows, but no timeframe was set. This version will require some changes for non-Windows platforms.
Ignore the post by @tudorserbanconnatix as he misunderstood something. An official version for .NET Core is not in the queue yet. So do not expect official versions for .NET Standard or .NET Core soon.

**Comment by ronnyek** [comment]
would you guys be opposed to pull requests for some or all of this functionality?

**Comment by ThomasHoevel** [comment]
@ronnyek We do not have SOPs for PRs. I would let my boss decide about any PR that is not just a bug fix.
PRs that focus on one feature will more likely be incorporated than PRs that add undocumented features, change the .NET version from 2 to 4.6, and change many members from internal to public without any documentation.
When creating a PR, make sure it fits in well with the folder and project structure that PDFsharp and MigraDoc currently have.

**Comment by andymac4182** [comment]
@ThomasHoevel It would be great if you could provide some guidelines for anyone wanting to get .net core support into PDFSharp since there is a large push from Microsoft and the community to Netstandard 2.0. It is possible to support back to 2 with the new multi-targetting support in the new csproj format. 

Personally I love this project and it helped heaps with some work we were doing and would love to be able to give back by helping update this project to dotnet core but need some idea if this work will be able to be merged back in or if moving to one of the forks is the only solution.

**Comment by ThomasHoevel** [comment]
@andymac4182 I don't know enough about the inner workings of PDFsharp to provide guidelines. I'm familiar with the image handling, but not with the font handling.

**Comment by YetaWF** [comment]
I just needed PDFSharp to render my PDFs as I transitioned to ASP.NET Core 2.0.
I forked PDFSharp and added the support (I needed). It's not pretty, but it will work until the official PDFSharp library is updated.
https://github.com/YetaWF/PDFsharp

**Comment by markusschaber** [comment]
@YetaWF The steps you described there look like the code itself actually works, and it's just the packaging which is missing (at least for your use case).

**Comment by YetaWF** [comment]
That's pretty much it. By using the prerelease version of System.Drawing.Common things work without changes, except for some WPF stuff.

**Comment by ststeiger** [comment]
If anybody needs this: I have it working WITHOUT System.Drawing (using Sixlabors instead). 
https://github.com/ststeiger/PdfSharpCore

**Comment by jamiehowarth0** [comment]
> If anybody needs this: I have it working WITHOUT System.Drawing (using Sixlabors instead).
> https://github.com/ststeiger/PdfSharpCore

I've just sent a pull request to use .NET Core 2.1 LTS today for your version. Can we make this into a Nuget package to get out there?

**Comment by ststeiger** [comment]
@benjaminhowarth1: I see no pull request anywhere. 

There already is a nuget package:
https://www.nuget.org/packages/PdfSharpCore/
It's maintained by [startnow65](https://github.com/startnow65).

Also, you might be interested in the full current version of PdfSharp for NetStandard 2.0 with System.Drawing.Common:
https://github.com/ststeiger/PdfSharpNetStandard

**Comment by jamiehowarth0** [comment]
Sorry I closed it - realised more work was needed on the compilation constants

**Comment by guylando** [comment]
Any of the suggested libraries is the official solution for net core? official in the sense that its supported by the same company supporting http://pdfsharp.net/

**Comment by TH-Soft** [comment]
> Any of the suggested libraries is the official solution for net core?
No, definitely not. As of now there is no "official" PDFsharp version for .NET Core.

**Comment by guylando** [comment]
@TH-Soft thanks for the answer

@ststeiger in https://github.com/ststeiger/PdfSharpCore you wrote that its a PARTIAL port of pdfsharp.xamarin. Can it be used as a drop-in replacement of PDFsharp-MigraDoc-gdi package for net core projects? Also what is the practical difference between it and your net standard pdfsharp port for the developer who uses it? 

In this thread there are more than 3 different ports proposed so its not clear to decide which one to use

**Comment by YetaWF** [comment]
FYI, I noticed that the official PdfSharp nuget package works fine in ASP.NET Core 2.2. This was not true in prior versions of ASP.NET Core. This means that with 2.2 none of these ports are required, just use the original package.

**Comment by guylando** [comment]
Interesting, I now see that net core developers worked on porting System.Drawing:
https://github.com/dotnet/corefx/issues/20325
https://developers.de/2018/01/22/how-to-use-system-drawing-in-net-core/
Also there is windows compatability pack having System.Drawing inside:
https://blogs.msdn.microsoft.com/dotnet/2017/11/16/announcing-the-windows-compatibility-pack-for-net-core/
And in net core 3.0 it will be FULLY supported: https://github.com/dotnet/corefx/pull/33092

**Comment by ststeiger** [comment]
@guylando: 
https://github.com/ststeiger/PdfSharpCore
is a port for .NET Core 1.6+
It doesn't use System.Drawing.dll. 
It is based on PdfSharp 1.32 and has some missing and some half-working features.

https://github.com/ststeiger/PdfSharpNetStandard
Is for NetStandard 2.1+, and is based on PdfSharp 1.5. 
Very few things are missing (https://github.com/ststeiger/PdfSharpNetStandard/tree/master/PdfSharp_Removed https://github.com/ststeiger/PdfSharpNetStandard/tree/master/MigraDoc_Rendering_Removed). 
PdfSharpNetStandard should work work just like PdfSharp. 

PdfSharpCore has changed PdfSharp to be based on the .NET-Core High-Performance implementations. 
It should have a better/higher performance - in theory. 

If you have mono installed (libgdiplus.so) then System.Drawing should work on Linux, too (not sure if this is true).

**Comment by brentboden1** [comment]
Is there any known date when we can expect an official release of pdfsharp for .net standard ?

**Comment by ThomasHoevel** [comment]
An official version for .NET Core is not in the queue yet. So do not expect official versions for .NET Standard or .NET Core soon.

**Comment by guylando** [comment]
@ThomasHoevel Asp.net core 3 does not compile to .Net framework which would prevent using pdfsharp

**Comment by ststeiger** [comment]
@guylando: ASP.NET **Core** will be **.NET Framework** 5.0. 
There will be nothing else anymore afterwards. 
Only ASP.NET Core < 3 can be used with FULL .NET Framework < 5. 

https://devblogs.microsoft.com/dotnet/introducing-net-5/
https://devblogs.microsoft.com/dotnet/net-core-is-the-future-of-net/

> .NET Framework 4.8 will be the last major version of .NET Framework. If you have existing .NET Framework applications that you are maintaining, there is no need to move these applications to .NET Core. We will continue to both service and support .NET Framework, which includes bug–, reliability– and security fixes. It will continue to ship with Windows (much of Windows depends on .NET Framework) and we will continue to improve the tooling support for .NET in Visual Studio (Visual Studio is written on .NET Framework).

**Comment by Methuselah96** [comment]
@ststeiger That's not entirely accurate. ASP.NET Core will (eventually) be .NET 5, which is not .NET Framework, it's just the next version of .NET Core that's being re-branded.

When ASP.NET Core uses .NET 5, it will still be able to depend on packages that use .NET Standard <=2.1. So in order to be compatible with the most number of frameworks, this library [should be ported to .NET Standard 2.0](https://docs.microsoft.com/en-us/dotnet/standard/net-standard#net-implementation-support). Then this library would support existing applications that are already written .NET Framework >=4.6.1 and also support new applications that will be written in .NET Core >=2.0 as well as .NET 5 when it comes out.

**Comment by B2alex** [comment]
This is a long chain, just wanted to check in and see if what I am understanding is correct.

PdfSharpCore
https://github.com/ststeiger/PdfSharpCore

Is the only port that supports .net core and got off of System.Drawing, for in favor of ImageSharp?

Also, has anyone been successful in getting a GDI based port to work on a linux container?

**Comment by flipchart** [comment]
@B2alex  We use it in production on a docker container running under Ubuntu linux on Azure. We modified to get it working and packaged it on nuget here https://www.nuget.org/packages/Drive.MigraDocPdfSharp

Basic code to get things running

```csharp
using System;
using System.Collections.Generic;
using System.IO;
using MigraDoc.DocumentObjectModel;
using MigraDoc.Rendering;
using PdfSharp.Drawing;

namespace PdfEg
{
    internal class PdfExample
    {
        private void RunPdfExample()
        {
            var fonts = new List<string>();

            if (Environment.OSVersion.Platform == PlatformID.Unix)
            {
                fonts.AddRange(Directory.EnumerateFiles("/usr/share/fonts/truetype/msttcorefonts"));
                fonts.AddRange(Directory.EnumerateFiles("/usr/share/fonts/truetype"));
            }
            else if (Environment.OSVersion.Platform == PlatformID.Win32NT)
            {
                fonts.AddRange(Directory.EnumerateFiles(Path.Combine(System.Environment.GetEnvironmentVariable("windir"), "Fonts")));
            }

            foreach (var font in fonts)
            {
                try
                {
                    XPrivateFontCollection.Add(File.ReadAllBytes(font));
                }
                catch
                {
                    // todo: log
                }
            }

            var document = new Document();
            var section = document.AddSection();

            document.Styles.Normal.Font = new Font("Arial", Unit.FromPoint(8));

            var h1 = document.AddStyle("h1", "Normal");
            h1.Font.Bold = true;
            h1.Font.Size = Unit.FromPoint(15);

            // add to page here :)

            var renderer = new PdfDocumentRenderer(true)
            {
                Document = document,
            };
            renderer.RenderDocument();

            using (var stream = new MemoryStream())
            {
                renderer.PdfDocument.Save(stream);

                var data = stream.ToArray();
            }
        }
    }
}

```

Dockerfile which we use. Note the `entrypoint.sh` file just runs the `dotnet` command on the main dll. Replace with your use

```dockerfile
FROM mcr.microsoft.com/dotnet/core/aspnet:2.2

ARG source=.

WORKDIR .
ENV ASPNETCORE_URLS http://*:5000
EXPOSE 5000/tcp

ARG NODEREPO="node_9.x"
ARG DISTRO="stretch"

RUN apt-get update -q && apt-get install -q -y \
 curl \
 apt-transport-https \
 apt-utils \
 dialog \
 gnupg

RUN curl -sSO https://deb.nodesource.com/gpgkey/nodesource.gpg.key
RUN apt-key add nodesource.gpg.key
RUN echo "deb https://deb.nodesource.com/${NODEREPO} ${DISTRO} main" > /etc/apt/sources.list.d/nodesource.list
RUN echo "deb-src https://deb.nodesource.com/${NODEREPO} ${DISTRO} main" >> /etc/apt/sources.list.d/nodesource.list

RUN (echo "deb http://http.debian.net/debian/ ${DISTRO} main contrib non-free" > /etc/apt/sources.list) && (echo "deb http://http.debian.net/debian ${DISTRO}-backports main" > /etc/apt/sources.list.d/backports.list) && (echo "deb http://http.debian.net/debian/ ${DISTRO}-updates main contrib non-free" >> /etc/apt/sources.list) && (echo "deb http://security.debian.org/ ${DISTRO}/updates main contrib non-free" >> /etc/apt/sources.list)
RUN echo ttf-mscorefonts-installer msttcorefonts/accepted-mscorefonts-eula select true | debconf-set-selections
RUN apt-get update && apt-get -y install \
  libgdiplus \
  ttf-mscorefonts-installer \
  nodejs
RUN cd /usr/lib && ln -s libgdiplus.so gdiplus.dll
RUN ln -s /lib/x86_64-linux-gnu/libdl.so.2 /lib/libdl.so
RUN ln -s /usr/lib/x86_64-linux-gnu/libX11.so.6 /lib/libX11.so

COPY $source .

RUN ["chmod", "+x", "/entrypoint.sh"]

ENTRYPOINT ["/entrypoint.sh"]

```

**Comment by jespersh** [comment]
The TFM concepts `netstandard` and `netcoreapp` has been merged into the `net5.0` and its platform specific TargetFramework `net5.0-windows`, `net5.0-watchos`, `net5.0-tvos`, `net5.0-macos`, `net5.0-ios`, `net5.0-android`

**Comment by flensrocker** [comment]
But with a target of `net5.0` you can't use that package in classic .Net Framework projects. `netstandard2.0` would be the most flexible one. And Pdfsharp doesn't use anything outside of .Net Standard 2.0.

**Comment by ststeiger** [comment]
Agree with @flensrocker. 
Anything but NETSTANDARD_2_0 is pointless.

**Comment by ThomasHoevel** [comment]
As of PDFsharp 6.0.0, PDFsharp is now using .NET 6.
